upgrade-2.x.sh

```diff
@@ -249,6 +250,176 @@ function device_type_match {
+
+    case "${SLUG}" in
+        raspberry*)
+            log "Running pre-update fixes for ${SLUG}"
+            pre_update_pi_bootfiles_removal
+            ;;
+        *)
+            log "No device-specific pre-update fix for ${SLUG}"
+    esac
+
+    docker pull "${update_package}" || log ERROR "Couldn't pull docker image..."
+    mkfifo /tmp/resinos-image.docker
+    docker save "${update_package}" > /tmp/resinos-image.docker &
+    mkdir -p /mnt/data/resinhup/tmp
+
+    log "Starting hostapp-update"
+    # Note that the following docker daemon is started with a different --bip and --fixed-cidr
+    # setting, otherwise it is clashing with the system docker on resinOS >=2.3.0 || <2.5.1
+    # and then docker pull would not succeed
+    # The requrired volumes are:
+    #  * /mnt/sysroot/inactive: the new rootfs
+    #  * /mnt/boot: the boot partition
+    #  * /tmp/resinos-image: the FIFO file that is used to transfer the resinOS image into the container
+    #  * /mnt/data/resinhup/tmp: a location to use as docker temporary directory within the container
+    # shellcheck disable=SC2016
+    docker run \
+      --rm \
+      --name resinhup \
+      --privileged \
+      -v "${temp_sysroot}":/mnt/sysroot/inactive \
+      -v /mnt/boot:/mnt/boot \
+      -v /tmp/resinos-image.docker:/resinos-image.docker \
+      -v /mnt/data/resinhup/tmp:/mnt/data/resinhup/tmp \
+      "${update_package}" \
+      /bin/bash -c 'storage_driver=$(cat /boot/storage-driver) ; DOCKER_TMPDIR=/mnt/data/resinhup/tmp/ dockerd --log-driver=journald -s=$storage_driver --data-root='"${sysroot}"'/docker -H unix:///var/run/docker-host.sock --pidfile=/var/run/docker-host.pid --exec-root=/var/run/docker-host --bip 10.114.201.1/24 --fixed-cidr=10.114.201.128/25 --iptables=false & sleep 10; echo "Starting hostapp-update"; hostapp-update -f /resinos-image.docker' \
+    || log ERROR "Update based on hostapp-update has failed..."
+}
+
+function find_partitions {
+    # Find which partition is / and which we should write the update to
+    root_part=$(findmnt -n --raw --evaluate --output=source /)
+    log "Found root at ${root_part}..."
+    case $root_part in
+        *2)
+            # on 2.x the following device types have these kinds of results for $root_part
+            # raspberrypi: /dev/mmcblk0p2
+            # beaglebone: /dev/disk/by-partuuid/93956da0-02
+            # NUC: /dev/sda2
+            root_dev=${root_part%2}
+            update_part=${root_dev}3
+            update_part_no=3
+            update_label=resin-rootB
+            ;;
+        *3)
+            root_dev=${root_part%3}
+            update_part=${root_dev}2
+            update_part_no=2
+            update_label=resin-rootA
+            ;;
+        *rootA)
+            old_label=resin-rootA
+            update_label=resin-rootB
+            root_part_dev=$(blkid | grep "${old_label}" | awk '{print $1}' | sed 's/://')
```

Comment:
Can we do this with `findmnt` or with `$(readlink -f /dev/disk/by-label/${old_label})`?

Reply:
Updated.